<a href="https://colab.research.google.com/github/piaseckazaneta/Python/blob/main/Mini_projekt_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

WSTĘP

Tutaj umieścić opis projektu, źródła danych i cel projektu
 - dokleić tu wszystko co jest w opisie projektu głównego, ile będzie zadań (małych celów) i że będzie na koniec mapa

1. IMPORT BIBLIOTEK

In [ ]:
import pandas as pd

from google.colab import drive

drive.mount("/content/drive")



Mounted at /content/drive


2. WCZYTANIE DANYCH

Dane pobierane są z dysku google, przez co nie ma potrzeby wczytywania ich przy każdym uruchomieniu kodu.



In [ ]:
# Ścieżki do dysku google
storm_events_details = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_details_2025.csv"
storm_events_locations = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_locations_2025.csv"
storm_events_fatalities = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_fatalities_2025.csv"

# Wczytanie danych
df_details = pd.read_csv(storm_events_details)
df_locations = pd.read_csv(storm_events_locations)
df_fatalities = pd.read_csv(storm_events_fatalities)


3. INSPEKCJA DANYCH

Analiza wzrokowa danych

In [ ]:
df_details.head()

,BEGIN_YEARMONTH,BEGIN_DAY,BEGIN_TIME,END_YEARMONTH,END_DAY,END_TIME,EPISODE_ID,EVENT_ID,STATE,STATE_FIPS,...,END_RANGE,END_AZIMUTH,END_LOCATION,BEGIN_LAT,BEGIN_LON,END_LAT,END_LON,EPISODE_NARRATIVE,EVENT_NARRATIVE,DATA_SOURCE
0,202503,31,1104,202503,31,1106,201366,1252415,GEORGIA,13,...,2.22,W,TYUS,33.4757,-85.238,33.4757,-85.238,A cold-front initiated a line of thunderstorms...,Tree down at the intersection of highway 5 and...,CSV
1,202503,30,1552,202503,30,1555,200337,1241136,MICHIGAN,26,...,1.47,NNE,EDWARDSBURG,41.7900,-86.100,41.8200,-86.070,A cold front pushed into the area during the a...,A brief EF-1 tornado was confirmed in Edwardsb...,CSV
2,202501,5,1800,202501,6,2227,197733,1222851,VIRGINIA,51,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure tracked across souther...,NaN,CSV
3,202501,3,1300,202501,3,1900,197761,1223112,MARYLAND,24,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure moved off into New Eng...,NaN,CSV
4,202501,3,1300,202501,3,1900,197761,1223113,MARYLAND,24,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure moved off into New Eng...,NaN,CSV


In [ ]:
df_fatalities.head()

,FAT_YEARMONTH,FAT_DAY,FAT_TIME,FATALITY_ID,EVENT_ID,FATALITY_TYPE,FATALITY_DATE,FATALITY_AGE,FATALITY_SEX,FATALITY_LOCATION
0,202512,4,0,60389,1296292,I,12/04/2025 00:00:00,40.0,M,Vehicle/Towed Trailer
1,202512,1,0,60391,1296789,I,12/01/2025 00:00:00,35.0,F,Boating
2,202512,3,0,60393,1297305,D,12/03/2025 00:00:00,32.0,M,In Water
3,202512,25,0,60553,1298352,I,12/25/2025 00:00:00,75.0,F,Vehicle/Towed Trailer
4,202512,26,0,60562,1298509,D,12/26/2025 00:00:00,44.0,M,In Water


In [ ]:
df_locations.head()

,YEARMONTH,EPISODE_ID,EVENT_ID,LOCATION_INDEX,RANGE,AZIMUTH,LOCATION,LATITUDE,LONGITUDE,LAT2,LON2
0,202508,204641,1287936,2,3.54,SSE,SINKING COVE,35.0015,-85.9601,3590,8557606
1,202508,203789,1269348,1,2.90,E,BLEASES XRDS,34.0440,-81.7699,342640,8146194
2,202508,203789,1269349,1,0.80,WSW,DOUGLAS,34.6637,-80.8118,3439822,8048708
3,202508,204597,1288676,4,1.45,WSW,HUNTSVILLE,34.7237,-86.6444,3443422,8638664
4,202508,204597,1288677,1,1.79,NW,FT PAYNE ISBELL ARPT,34.4900,-85.7400,3429400,8544400


**Komentarz**:
Tabele wyglądają na uporządkowane.
Poprawnie wyglądają formaty dat, skrótowe oznaczenia kierunków świata

Wyświetlenie wszystkich atrybutów każdej tabeli

In [ ]:
df_details.columns.to_list()

['BEGIN_YEARMONTH',
 'BEGIN_DAY',
 'BEGIN_TIME',
 'END_YEARMONTH',
 'END_DAY',
 'END_TIME',
 'EPISODE_ID',
 'EVENT_ID',
 'STATE',
 'STATE_FIPS',
 'YEAR',
 'MONTH_NAME',
 'EVENT_TYPE',
 'CZ_TYPE',
 'CZ_FIPS',
 'CZ_NAME',
 'WFO',
 'BEGIN_DATE_TIME',
 'CZ_TIMEZONE',
 'END_DATE_TIME',
 'INJURIES_DIRECT',
 'INJURIES_INDIRECT',
 'DEATHS_DIRECT',
 'DEATHS_INDIRECT',
 'DAMAGE_PROPERTY',
 'DAMAGE_CROPS',
 'SOURCE',
 'MAGNITUDE',
 'MAGNITUDE_TYPE',
 'FLOOD_CAUSE',
 'CATEGORY',
 'TOR_F_SCALE',
 'TOR_LENGTH',
 'TOR_WIDTH',
 'TOR_OTHER_WFO',
 'TOR_OTHER_CZ_STATE',
 'TOR_OTHER_CZ_FIPS',
 'TOR_OTHER_CZ_NAME',
 'BEGIN_RANGE',
 'BEGIN_AZIMUTH',
 'BEGIN_LOCATION',
 'END_RANGE',
 'END_AZIMUTH',
 'END_LOCATION',
 'BEGIN_LAT',
 'BEGIN_LON',
 'END_LAT',
 'END_LON',
 'EPISODE_NARRATIVE',
 'EVENT_NARRATIVE',
 'DATA_SOURCE']

Komentarz:
.head() ucięło kolumny w df_details.

Prezentacja nazw kolumn w postaci listy umożliwia zobaczenie wszystkich kolumn w tabeli.

Sprawdzenie rozmiarów tabel i typów kolumn

In [ ]:
df_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72241 entries, 0 to 72240
Data columns (total 51 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   BEGIN_YEARMONTH     72241 non-null  int64  
 1   BEGIN_DAY           72241 non-null  int64  
 2   BEGIN_TIME          72241 non-null  int64  
 3   END_YEARMONTH       72241 non-null  int64  
 4   END_DAY             72241 non-null  int64  
 5   END_TIME            72241 non-null  int64  
 6   EPISODE_ID          72241 non-null  int64  
 7   EVENT_ID            72241 non-null  int64  
 8   STATE               72241 non-null  object 
 9   STATE_FIPS          72241 non-null  int64  
 10  YEAR                72241 non-null  int64  
 11  MONTH_NAME          72241 non-null  object 
 12  EVENT_TYPE          72241 non-null  object 
 13  CZ_TYPE             72241 non-null  object 
 14  CZ_FIPS             72241 non-null  int64  
 15  CZ_NAME             72241 non-null  object 
 16  WFO 

Komentarz:

Użycie .info() służyło sprawdzeniu typów danych i potencjalnych braków danych dzięki kolumnie Non-Null Count.. Można odczytać też liczbę wierszy w tabeli.

Całkowita liczba wierszy: 72241.

X Daty ("date") są w formacie int64 - do konwersji.
X Kwoty ("damage") dotyczące zniszczeń są w formacie object - do konwersji.
X Skala tornada (tor_f_scale) jest w formacie object - do późniejszych obliczeń będzie potrzebny format liczbowy.

Braki w danych:
ok. 20 tys. z 72 tys. brakujących wartości w kolumnie "damage_crops" - do sprawdzenia (powinno być 0, czy zdarzenie nie dotyczy zniszczeń upraw?)
ok. 70 tys. brakujących wartości dla skali, długości i szerokości tornad (tor_f_scale, tor_length, tor_width) - do sprawdzenia, prawdopodobnie brakujące wiersze są innymi zdarzeniami niż tornado.

Reszta interesujących kolumn ma poprawny format i kompletną liczbę wartości.


In [ ]:
df_fatalities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 895 entries, 0 to 894
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   FAT_YEARMONTH      895 non-null    int64  
 1   FAT_DAY            895 non-null    int64  
 2   FAT_TIME           895 non-null    int64  
 3   FATALITY_ID        895 non-null    int64  
 4   EVENT_ID           895 non-null    int64  
 5   FATALITY_TYPE      895 non-null    object 
 6   FATALITY_DATE      895 non-null    object 
 7   FATALITY_AGE       743 non-null    float64
 8   FATALITY_SEX       785 non-null    object 
 9   FATALITY_LOCATION  895 non-null    object 
dtypes: float64(1), int64(5), object(4)
memory usage: 70.1+ KB


Komentarz:

Całkowita liczba wierszy: 895.
Wszystkie kolumny mają poprawny format danych.
W "fatality_age" i "fatality_sex" ok. 100 brakujących wartości - do sprawdzenia.

In [ ]:
df_locations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51870 entries, 0 to 51869
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   YEARMONTH       51870 non-null  int64  
 1   EPISODE_ID      51870 non-null  int64  
 2   EVENT_ID        51870 non-null  int64  
 3   LOCATION_INDEX  51870 non-null  int64  
 4   RANGE           51870 non-null  float64
 5   AZIMUTH         51870 non-null  object 
 6   LOCATION        51870 non-null  object 
 7   LATITUDE        51870 non-null  float64
 8   LONGITUDE       51870 non-null  float64
 9   LAT2            51870 non-null  int64  
 10  LON2            51870 non-null  int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 4.4+ MB


Komentarz:

Całkowita liczba wierszy: 51870.
Wszystkie kolumny mają poprawny format danych i 0 brakujących wartości.

Sprawdzenie pełnej listy zjawisk, jakie wystąpiły.

In [ ]:
df_details['EVENT_TYPE'].unique()

array(['Thunderstorm Wind', 'Tornado', 'Winter Storm', 'Winter Weather',
       'High Wind', 'Extreme Cold/Wind Chill', 'Hail', 'Flash Flood',
       'Blizzard', 'Strong Wind', 'Heavy Snow', 'Flood',
       'Marine High Wind', 'Excessive Heat', 'Marine Thunderstorm Wind',
       'Marine Hail', 'Frost/Freeze', 'Sneakerwave', 'Drought',
       'Funnel Cloud', 'Heavy Rain', 'Heat', 'Wildfire', 'Dust Storm',
       'Debris Flow', 'Cold/Wind Chill', 'Rip Current', 'Waterspout',
       'Coastal Flood', 'Lightning', 'Avalanche', 'High Surf',
       'Lake-Effect Snow', 'Dense Fog', 'Ice Storm', 'Sleet',
       'Astronomical Low Tide', 'Marine Strong Wind', 'Freezing Fog',
       'Marine Dense Fog', 'Marine Tropical Storm', 'Tropical Storm',
       'Tsunami', 'Seiche', 'Tropical Depression', 'Storm Surge/Tide',
       'Lakeshore Flood', 'Dust Devil'], dtype=object)

SELEKCJA ATRYBUTÓW

Zmiana nazw kolumn w df-ach na małe litery

In [ ]:
df_details.columns = df_details.columns.str.lower()
df_fatalities.columns = df_fatalities.columns.str.lower()
df_locations.columns = df_locations.columns.str.lower()

Selekcja zmiennych do analiz

In [ ]:
df_details_analysis = df_details[["event_id", "state", "event_type", "begin_date_time", "damage_property", "damage_crops", "tor_f_scale", "tor_length", "tor_width"]]

In [ ]:
df_fatalities_analysis = df_fatalities[["event_id", "fatality_age", "fatality_sex", "fatality_location"]]

In [ ]:
df_locations_analysis = df_locations[["event_id", "latitude", "longitude"]]

Sprawdzenie wartości w wyselekcjonowanych kolumnach

In [ ]:
df_details_analysis.head()

,event_id,state,event_type,begin_date_time,damage_property,damage_crops,tor_f_scale,tor_length,tor_width
0,1252415,GEORGIA,Thunderstorm Wind,31-MAR-25 11:04:00,1.00K,NaN,NaN,NaN,NaN
1,1241136,MICHIGAN,Tornado,30-MAR-25 15:52:00,100.00K,0.00K,EF1,2.59,100.0
2,1222851,VIRGINIA,Winter Storm,05-JAN-25 18:00:00,NaN,NaN,NaN,NaN,NaN
3,1223112,MARYLAND,Winter Weather,03-JAN-25 13:00:00,NaN,NaN,NaN,NaN,NaN
4,1223113,MARYLAND,Winter Weather,03-JAN-25 13:00:00,NaN,NaN,NaN,NaN,NaN


Wnioski dotyczące wartości w kolumnach tabeli:
- wartości zniszczeń ("damage property" i "damage crops" są tekstowe -> należy je zamienić na wartości numeryczne w celu późniejszego oszacowania zniszczeń
- wartości w kolumnie "begin_date_time" zawierają date i czas -> należy te informacje rozdzielić na osobne kolumny, aby móc później określić sezonowość (miesiąc) i porę dnia (godzina) poszczególnych zjawisk
- "tor_f_scale" - Standaryzacja skali tornad: Oczyszczenie kolumny tor_f_scale (wyciągnięcie samej cyfry ze stringów typu "EF2").


KONWERSJA DANYCH

Cel: Naprawa i ujednolicenie formatów

Konwersja dat

Konwersja kolumny "begin_date_time" z df_details_analysis

In [ ]:
df_details_analysis["begin_date_time"] = pd.to_datetime(df_details_analysis["begin_date_time"])

/tmp/ipykernel_20186/2334216105.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_details_analysis["begin_date_time"] = pd.to_datetime(df_details_analysis["begin_date_time"])
/tmp/ipykernel_20186/2334216105.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_details_analysis["begin_date_time"] = pd.to_datetime(df_details_analysis["begin_date_time"])


Wyciągnięcie z daty nowych kolumn: month (do analizy sezonowej) oraz hour (do analizy porowej)

In [ ]:
df_details_analysis["begin_date_time"]

,begin_date_time
0,2025-03-31 11:04:00
1,2025-03-30 15:52:00
2,2025-01-05 18:00:00
3,2025-01-03 13:00:00
4,2025-01-03 13:00:00
...,...
72236,2025-09-16 14:53:00
72237,2025-09-16 14:54:00
72238,2025-09-09 16:58:00
72239,2025-09-09 19:32:00


In [ ]:
month = df_details_analysis["begin_date_time"].dt.month

In [ ]:
print(month)

0        3
1        3
2        1
3        1
4        1
        ..
72236    9
72237    9
72238    9
72239    9
72240    9
Name: begin_date_time, Length: 72241, dtype: int32


In [ ]:
hour = df_details_analysis["begin_date_time"].dt.hour

In [ ]:
print(hour)

0        11
1        15
2        18
3        13
4        13
         ..
72236    14
72237    14
72238    16
72239    19
72240    18
Name: begin_date_time, Length: 72241, dtype: int32


Funkcja zamieniająca tekstowe wartości zniszczeń (np. 10.00K, 2.50M) na czyste liczby typu float ($10000$, $2500000$) dla kolumn "damage_crops" i "damage_property" w df_details_analysis

In [ ]:
df_details_analysis["damage_property"].unique()

array(['1.00K', '100.00K', nan, '0.00K', '10.00K', '50.00K', '25.00K',
       '5.00K', '15.00K', '20.00K', '9.00K', '3.00K', '2.00K', '30.00K',
       '4.00K', '6.00K', '0.50K', '2.50K', '1.50K', '1.10K', '18.50M',
       '17.80M', '3.40M', '8.00K', '55.00K', '75.00K', '3.50K', '40.00K',
       '35.00K', '0.60K', '1.00M', '5.00M', '2.50M', '0.20K', '0.30K',
       '200.00K', '250.00K', '150.00K', '260.00K', '700.00K', '10.50K',
       '600.00K', '900.00K', '7.00K', '12.00K', '300.00K', '11.00K',
       '2.25K', '49.50K', '0.10K', '500.00K', '160.00K', '70.00K',
       '60.00K', '45.00K', '7.50K', '0.25K', '80.00K', '13.00K', '0.75K',
       '140.00K', '3.90M', '90.00K', '350.00K', '14.00K', '1.20M',
       '12.50K', '27.00K', '0.01K', '0.00M', '18.00M', '33.00M', '56.00M',
       '120.00K', '0.40K', '125.00K', '5000.00K', '870.00K', '800.00K',
       '19.00K', '1.50M', '148.00K', '268.00K', '222.00K', '825.00K',
       '2.08M', '1.20K', '1.60K', '22.50K', '750.00K', '215.00K',
       '

In [ ]:
def damage_costs(damage_value):
  if pd.isna(damage_value):
    return 0
  elif "K" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000
    return damage_value
  elif "M" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000000
    return damage_value
  elif "B" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000000000
    return damage_value
  else:
    float(damage_value)

In [ ]:
df_details_analysis["damage_property"] = df_details_analysis["damage_property"].apply(damage_costs)

/tmp/ipykernel_20186/1084519902.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_details_analysis["damage_property"] = df_details_analysis["damage_property"].apply(damage_costs)


In [ ]:
print(df_details_analysis["damage_property"])

0          1000.0
1        100000.0
2             0.0
3             0.0
4             0.0
           ...   
72236         0.0
72237         0.0
72238         0.0
72239         0.0
72240      3000.0
Name: damage_property, Length: 72241, dtype: float64


In [ ]:
df_details_analysis["damage_crops"] = df_details_analysis["damage_crops"].apply(damage_costs)

/tmp/ipykernel_20186/4082403526.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_details_analysis["damage_crops"] = df_details_analysis["damage_crops"].apply(damage_costs)


In [ ]:
print(df_details_analysis["damage_crops"])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
72236    0.0
72237    0.0
72238    0.0
72239    0.0
72240    0.0
Name: damage_crops, Length: 72241, dtype: float64


Wyciągnięcie z kolumny "tor_f_scale" samej liczby (wartość wyrażona jest w skali Fujita, np. EF1). Wartość liczbowa będzie niezbędna do obliczeń zależności powierzchni zniszczeń od wzrostu siły tornada.

Sprawdzenie unikalnych wartości występujących w kolumnie "tor_f_scale"

In [ ]:
df_details_analysis["tor_f_scale"].unique()

array([nan, 'EF1', 'EF0', 'EF3', 'EFU', 'EF2', 'EF4', 'EF5'], dtype=object)

Funkcja wyodrębniająca cyfrę z wartości w kolumnie "tor_f_scale".
'EFU' jest oznaczeniem tornada, które przechodziło przez pole i nie spowodowało uszkodzeń budynków, zatem nie można mu przypisać oznaczenia w skali 1-5.

In [ ]:
def scale_number(scale):
  if pd.isna(scale):
    return 'NaN'
  elif "EFU" in scale:
    return 'NaN'
  elif "EF" in scale:
    scale = scale[2:]
    return scale
  else:
    return scale

Zastosowanie napisanej funkcji dla kolumny "tor_f_scale" i utworzenie nowej kolumny z samą cyfrą.

In [ ]:
df_details_analysis["tor_f_scale_number"] = df_details_analysis["tor_f_scale"].apply(scale_number)

/tmp/ipykernel_20186/2350290250.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_details_analysis["tor_f_scale_number"] = df_details_analysis["tor_f_scale"].apply(scale_number)


Sprawdzenie unikalnych wartości występujących w nowej kolumnie

In [ ]:
df_details_analysis["tor_f_scale_number"].unique()

array(['NaN', '1', '0', '3', '2', '4', '5'], dtype=object)

Zamiana wartości na float.

In [ ]:
df_details_analysis["tor_f_scale_number"] = (df_details_analysis["tor_f_scale_number"]).astype(float)

/tmp/ipykernel_20186/2942268197.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_details_analysis["tor_f_scale_number"] = (df_details_analysis["tor_f_scale_number"]).astype(float)
